File: 剑指offer/39.数组中出现次数超过一半的数字.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        votes = 0
        x = None
        for num in nums:
            if votes == 0:
                x = num

            if num == x:
                votes += 1
            else:
                votes -= 1
        return x

File: 剑指offer/39.数组中出现次数超过一半的数字.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums)
    {
        int ans = 0;
        int vote = 0;

        for (int num : nums) {
            if (vote == 0) {
                ans = num;
            }

            if (num == ans) {
                vote++;
            } else {
                vote--;
            }
        }

        return ans;
    }
};